packages/components/src/utils/hooks/use-prop-ref.ts

```diff
@@ -0,0 +1,52 @@
+/**
+ * External dependencies
+ */
+// eslint-disable-next-line no-restricted-imports
+import type { MutableRefObject } from 'react';
+
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useRef } from '@wordpress/element';
+
+/**
+ * Creates a reference for a prop. This is useful for preserving dependency
+ * memoization for hooks like useCallback.
+ *
+ * @example
+ * ```js
+ * // Referencing a simple prop, used in a useCallback function.
+ * const valueRef = usePropRef(value)
+ *
+ * const increment = useCallback(() => {
+ *   const value = valueRef.current
+ *   onChange(value + 1)
+ * }, [onChange, valueRef])
+ * ```
+ *
+ * ---
+ *
+ * Multiple props can be passed in using an `object`.
+ *
+ * @example
+ * ```js
+ * const propRefs = usePropRef({ value, step })
+ *
+ * const increment = useCallback(() => {
+ *   const { value, step } = propRefs.current
+ *   onChange(value + step)
+ * }, [onChange, propRefs])
+ * ```
+ *
+ * @param  prop
+ * @return The prop reference.
+ */
+export function usePropRef< T >( prop: T ): MutableRefObject< T > {
+	const propRef = useRef( prop );
+
+	useEffect( () => {
+		propRef.current = prop;
+	}, [ prop ] );
+
+	return propRef;
+}
```

Comment:
I would call it `useLiveRef` or `useLatestRef` and maybe change the return type to `RefObject` instead of `MutableRefObject` as we don't really want to allow mutations outside of this hook.

We can also use `useLayoutEffect` instead of `useEffect` so we get the updated value in other `useLayoutEffect`'s. Also, the prop dependency seems unnecessary as we're just assigning a value.
```suggestion
export function useLatestRef< T >( prop: T ): RefObject< T > {
	const propRef = useRef( prop );

	useLayoutEffect( () => {
		propRef.current = prop;
	} );

	return propRef;
}
```